src/node/internal/http_request.js

```diff
@@ -0,0 +1,178 @@
+import {inherits} from 'node-internal:http_util';
+import {IncomingMessage, readyStates as rStates} from 'node-internal:http_response';
+import {Writable} from 'node-internal:streams_writable';
+import {Buffer} from 'node-internal:internal_buffer';
+
+export default ClientRequest;
+
+function ClientRequest(opts) {
+  var self = this
+  Writable.call(self)
+
+  self._opts = opts
+  self._body = []
+  self._headers = {}
+  if (opts.auth)
+    self.setHeader('Authorization', 'Basic ' + new Buffer(opts.auth).toString('base64'))
+  Object.keys(opts.headers).forEach(function(name) {
+    self.setHeader(name, opts.headers[name])
+  })
+
+  self._mode = 'fetch'
+
+  self.on('finish', function() {
+    self._onFinish()
+  })
+}
+
+inherits(ClientRequest, Writable)
+// Taken from http://www.w3.org/TR/XMLHttpRequest/#the-setrequestheader%28%29-method
+var unsafeHeaders = [
+  'accept-charset',
+  'accept-encoding',
+  'access-control-request-headers',
+  'access-control-request-method',
+  'connection',
+  'content-length',
+  'cookie',
+  'cookie2',
+  'date',
+  'dnt',
+  'expect',
+  'host',
+  'keep-alive',
+  'origin',
+  'referer',
+  'te',
+  'trailer',
+  'transfer-encoding',
+  'upgrade',
+  'user-agent',
+  'via'
+]
+ClientRequest.prototype.setHeader = function(name, value) {
+  var self = this
+  var lowerName = name.toLowerCase()
+    // This check is not necessary, but it prevents warnings from browsers about setting unsafe
+    // headers. To be honest I'm not entirely sure hiding these warnings is a good thing, but
+    // http-browserify did it, so I will too.
```

Comment:
This is all from a polyfill then? I'm not sure this is the way we want to go. My intention is to port the Node.js implementation directly and not start from a polyfill for the http pieces.